warehouse/forklift/legacy.py

```diff
@@ -519,6 +516,19 @@ def file_upload(request):
                     project_help=request.help_url(_anchor="verified-email"),
                 ),
             ) from None
+        # Ensure user has enabled 2FA before they can upload a file.
+        if not request.user.has_two_factor:
```

Comment:
I don't think this can happen in this case because of our restrictions on Trusted Publishing, but just calling it out: `request.user` can be `None` if the uploader is a Trusted Publisher instead of a user-controlled API token.

Reply:
Thanks! I think this is already guarded by `if request.user:` above, isn't it? https://github.com/pypi/warehouse/pull/16365/files#diff-bffd49e4f943217b54b4cfcf5835737bec7a145360bd1dc2ec1541d2dcafad16L501-L503

Reply:
Ah yep, indeed -- I missed that in the diff context 😅